### PR TITLE
[Merged by Bors] - feat(SetTheory): definition of initial ordinals, ω₁ as an ordinal, ordinal-indexed unions

### PR DESCRIPTION
**Pull request by sterraf**
- I setup notation for the first ordinal in each cardinality.
- `ω₁` is defined as an ordinal, not as an `out` (cf. `MeasureTheory.CardMeasurableSpace`).
- Lemma using the cofinality of `ω₁`.
- Lemma on the cardinality of ordinal-indexed `iUnion`s in preparation for material on the Borel hierarchy.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by sterraf** [comment]
I'm trying to revive a [mathlib3 PR](https://github.com/leanprover-community/mathlib/pull/17972) of mine in parts. 

The first one ([Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Defining.20initial.20ordinals/near/338425902)) only involves ordinals and cardinals.

In the old PR, @vihdzp contributed with golfing a proof, and @sgouezel and @alreadydone help me with style during the review process. I should probably have to add (some of) them as coauthors. Nevertheless, I ported this to Lean4 by hand and I might have missed some of the new style conventions.

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Cardinal/Ordinal.lean:1488]: golf
```suggestion
lemma omega_lt_omega1 : ω < ω₁ := ord_aleph0.symm.trans_lt (ord_lt_ord.mpr (aleph0_lt_aleph_one))
```

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Cardinal/Ordinal.lean:1519]: mathlib4 style + golf
```suggestion
    (hi : i ≤ κ.ord) (hκ : ℵ₀ ≤ κ) (A : Ordinal → Set β)
    (hA : ∀ j < i, #(A j) ≤ κ) :
    #(⋃ j < i, A j) ≤ κ := by
  simp_rw [← mem_Iio, biUnion_eq_iUnion, iUnion, iSup, ← i.enumIsoOut.symm.surjective.range_comp]
```
I would suggest also use the names `c` and `o` here. It's natural to use `i : Ordinal` above where it serves as an index, but not so much here.
Also suggest changing `hi : i ≤ κ.ord` to `hi : i.card ≤ κ` to makes it more general.

**Review by alreadydone — COMMENTED** [review]
LGTM otherwise

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (sterraf) [reply Mathlib/SetTheory/Ordinal/Basic.lean:1392]: Should I add that `_iff`?

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (sterraf) [reply Mathlib/SetTheory/Ordinal/Basic.lean:1391]: You mean the *variable names* in the docstring? Sorry to be pedantic, there were already some comments on the (older) docstring.

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Ordinal/Basic.lean:1393]: This is what I meant: change the variable names in both the statement and the docstring.

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Cardinal/Ordinal.lean:1511]: Variable names here haven't been changed; do you agree it's better to use `c` and `o` here following the rest of the file?

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (alreadydone) [reply Mathlib/SetTheory/Ordinal/Basic.lean]: You're welcome to! It may be better to derive this from that stronger result.

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (sterraf) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean:1511]: I'm actually following your suggestions without hesitations---i.e., I agree completely. 

I just added all the proposed commits in one batch and only later I realized that it did not pass CI. (I did not--and do not--know how to make a more comprehensive commit later and add you as a coauthor.)
Also, I realized I didn't remove the `awaiting-review` label before I double check the stuff. 

Btw, did you see Mario's [comments](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.236404.20defining.20initial.20ordinals/near/382512001)? I was a bit concerned but perhaps we can go ahead with this.

Comment (sterraf) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean:1511]: > I'm actually following your suggestions without hesitations---i.e., I agree completely.
> 
> I just added all the proposed commits in one batch and only later I realized that it did not pass CI. (I did not--and do not--know how to make a more comprehensive commit later and add you as a coauthor.) 

My plan was to complete the variable renaming and indent styling after committing the first batch.

Comment (alreadydone) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean:1511]: > I just added all the proposed commits in one batch and only later I realized that it did not pass CI.

No worries!

> I did not--and do not--know how to make a more comprehensive commit later and add you as a coauthor.

Normally you'd just commit yourself and add me in a Co-authored-by line in the PR description.

> Btw, did you see Mario's [comments](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.236404.20defining.20initial.20ordinals/near/382512001)? I was a bit concerned but perhaps we can go ahead with this.

I actually couldn't find where the range of `Cardinal.ord` is used to define initial ordinals; do you know?

Comment (sterraf) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean]: > Normally you'd just commit yourself and add me in a Co-authored-by line in the PR description.
>

Done
 
> > Btw, did you see Mario's [comments](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.236404.20defining.20initial.20ordinals/near/382512001)? I was a bit concerned but perhaps we can go ahead with this.
> 
> I actually couldn't find where the range of `Cardinal.ord` is used to define initial ordinals; do you know?

No, I don't, and I guess it's not. I think he meant that adding this particular def was not necessary, or that adding it as an extra layer would be problematic from an API point of view

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (sterraf) [reply Mathlib/SetTheory/Ordinal/Basic.lean]: > You're welcome to! You just need to combine `Order.lt_succ_iff` and `Cardinal.lt_ord`, I think. It may be better to derive `card_le_of_le_ord` from that stronger result.

I added this `card_le_iff`, but I'm not sure how to derive the older result from this (the present proof is rather short; and the statement, simple enough).

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Cardinal/Ordinal.lean:1504]: Did you forget to change κ to c here?

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Ordinal/Basic.lean:1396]: You can do this but it's longer and depends on `ord_lt_ord` which appears later, so forget about it.
```suggestion
    o.card ≤ c :=
  card_le_iff.2 (ho.trans_lt <| ord_lt_ord.2 <| lt_succ c)
```

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (alreadydone) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean]: > Normally you'd just commit yourself and add me in a Co-authored-by line in the PR description.

No longer to do this since you already commited my suggestions (and I also commited). I've removed the line because otherwise my name will appear twice.

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (sterraf) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean:1504]: :man_facepalming:

**Review by alreadydone — APPROVED** [review]
Thanks!

maintainer merge

File: Mathlib/SetTheory/Cardinal/Cofinality.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Cardinal/Cofinality.lean:1252]: It may feel a bit redundant to enclose a `namespace ... end` with a `section`, but maybe it's a good way to organize things. (Personally I might just add a comment rather than using a section; sections can be collapsed in VSCode but you may as well collapse the namespace.)

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Ordinal/Basic.lean:1390]: This should probably just say
```suggestion
initial ordinal of cardinality `c`, then its cardinal is no greater than `c`.
```
so maybe delegate so we can fix this before merge.

File: Mathlib/SetTheory/Cardinal/Cofinality.lean
Comment (sterraf) [reply Mathlib/SetTheory/Cardinal/Cofinality.lean:1252]: The style I have seen is having a section for each new concept, I was only trying to mimic this. As always, I open to modification if it is better for mathlib.

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (alreadydone) [inline Mathlib/SetTheory/Cardinal/Ordinal.lean]: These are some interesting lemmas but I'm also fine with it if you don't want to keep them.

**Review by alreadydone — APPROVED** [review]

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (sterraf) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean]: It seems that the first 3 are indeed redundant once the definition layer was stripped, but perhaps the last one is worth keeping. I'll put that one back.

File: Mathlib/SetTheory/Cardinal/Ordinal.lean
Comment (digama0) [inline Mathlib/SetTheory/Cardinal/Ordinal.lean:1459]:
```suggestion
scoped notation "ω_" o => ord <| aleph o
```
this adds an extra beta redex in uses of the notation, plus using `ω_` without applying it to something looks weird and confusing

Comment (sterraf) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean:1459]: For the first thing, that's simply me not knowing how to use the `notation` command, sorry for that.

For the notation itself, I do not like `ω_` either; it was a compromise solution [discussed on Zulip](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Defining.20initial.20ordinals/near/338425902). I would have liked to have a "subscript argument", but from that discussion I understood that it is not possible.

Do you suggest that the `ωₙ` notation be introduced on demand?

Comment (alreadydone) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean:1459]: Let's commit the suggestion and call it a day :) I think Mario is saying that a standalone `ω_` looks weird but `ω_ o` looks okay. I think it's definitely reasonable to introduce local notations `ω₁`, `ω₂` etc. to make things more readable, but if you want to use the `Ordinal.initial` function rather than just one of its values then you can't possibly fit an arbitrary expression (or even character) into the subscript, and would have to write `ω_ expr`.

Comment (sterraf) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean:1459]: Of course I accept the suggestion, I honestly wanted to know if it was deemed even better not to introduce `ω_` at all.

**Comment by kmill** [comment]
What's the state of this PR? I see "awaiting-review" has been taken off -- does that mean you don't want it to be merged yet @sterraf ?

I'll delegate merging to @alreadydone 

bors d=alreadydone

**Comment by sterraf** [comment]
Hi,  I removed the label as Junyan had already approved it, and Mario took
a look.

I was just waiting to see if Mario had more comments, but I'd be happy to
have it merged. Thanks!

PST.-

—sent from mobile

**Comment by kmill** [comment]
Ah, ok, just so you know we use "awaiting-review" to signal "the author approves of merging", so your removing it after approval could have meant "the author does not want this merged right now." No need to remove the label if you get reviewer approval.

Thanks!

bors r+

**Comment by vihdzp** [comment]
Do you have any plans to continue porting the Borel hierarchy? That'd be really nice to have!

**Comment by sterraf** [comment]
> Do you have any plans to continue porting the Borel hierarchy? That'd be really nice to have!

Hi! I'm a bit disconnected from this, and also recall some new contributors were working on Descriptive Set Theory. If the [proposed definition at mathlib3](https://github.com/leanprover-community/mathlib3/pull/17972) is still sensible and does not interfere with (or better, is of interest to) them, I might find some time to do it in January.

**Pull request closed without merging**
